### DISCUSSION TITLE
why does the root element in JsonConfigurationFileParser must be an object?

### DISCUSSION THREAD

**Discussion by yoziv** [dotnet/runtime#56638] [Q&A] [answered]
i am looking at the JsonConfigurationFileParser class specifically at the parse stream function [(ParseStream)](https://github.com/dotnet/runtime/blob/b2e483e3d500989cac05c7faf2122560b5ee9e9b/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationFileParser.cs#L22) in this specific section:

`if (doc.RootElement.ValueKind != JsonValueKind.Object)
{
                    throw new FormatException(SR.Format(SR.Error_InvalidTopLevelJSONElement, doc.RootElement.ValueKind));
                }`
i see you only allow the root element of a json to be an object although a json where the root element is an array is also a valid json for example: 
`[
  {
    "version": 1,
    "handlerEnvironment": {
      "logFolder": ".\\logs",
      "configFolder": ".\\RuntimeSettings",
      "statusFolder": ".\\Status",
      "heartbeatFile": ".\\Status\\HeartBeat.Json"
    }
  }
]` 
why does that restriction exist ? and how can i use this json file as a config ( i don't control the json file so i can't change it)

Thanks,
Yotam

**Comment by LeaFrock** [comment, marked as answer]
> why does that restriction exist ?

It's just a convention of design. As the codes(`VisitElement`, `VisitValue`) below the line you show informs, the parser accepts the premise that the json content should be a json object rather than other json types.

> how can i use this json file as a config

Though you cannot control the files, you can check and rewrite the content of each json file into a new one, then let the parser use it. 
``` json
{"MySettings":[ { "version": 1, "handlerEnvironment": { "logFolder": ".\\logs", "configFolder": ".\\RuntimeSettings", "statusFolder": ".\\Status", "heartbeatFile": ".\\Status\\HeartBeat.Json" } } ]}
```
Or you may have to write your own parser and replace the default one in DI.

**Reply by davidfowl** [reply]
Just write your own configuration provider.

**Reply by LeaFrock** [reply]
Oh yep... :laughing:

**Reply by yoziv** [reply]
already did ☺ just wondering what is the reason for the premise

**Reply by davidfowl** [reply]
It wouldn't be intuitive with how the configuration values are accessed. When we encounter a JSON array the access looks like:

```JSON
[
  {
    "version": 1,
    "handlerEnvironment": {
      "logFolder": ".\\logs",
      "configFolder": ".\\RuntimeSettings",
      "statusFolder": ".\\Status",
      "heartbeatFile": ".\\Status\\HeartBeat.Json"
    }
  }
]
```

```C#
var version = Configuration["0:version"];
var logFolder = Configuration["0:handlerEnvironment:logFolder"];
```

It could absolutely be supported, its just not really mainline scenario we've ever seen anyone complain about.

**Reply by LeaFrock** [reply]
Depend on what is the most common. And leave another way you can define by self (your own configuration provider).

**Reply by yoziv** [reply]
Thanks guys :)
